### ISSUE TITLE
Prioritize the use of local peers 

### ISSUE THREAD

**Issue by fredy314** [anacrolix/torrent#894]
If another torrent client is installed on the same server as TorrServer. in which the same torrent is added and the file is already completely downloaded, then I believe it will work much faster if peers on 127.0.0.1 or the local network are taken first to watch the video

**Comment by anacrolix** [comment]
I don't think LSD is implemented, but it could be possible to preference local IPs. How are you adding those? I suspect having private IPs returned from trackers won't be handled well currently.

**Comment by philrhc** [comment]
Interestingly this was researched in this paper https://arxiv.org/abs/2507.20116, II-A Traditional P2P Approaches:
> even when local peers were available in LAN 2, Kraken still fetched approximately 10% of the image blocks from remote peers in LAN 1. Crucially, this small fraction consumed over 95% of the inter-LAN bandwidth. This behavior stems from BitTorrent’s peer selection strategy, which prioritizes global peer diversity and piece availability over network locality. As a result, the narrow uplink becomes saturated, delaying downloads for all nodes and negating the latency advantage of local caching.

**Comment by anacrolix** [comment]
Hm yeah, I bet it's nontrivial to factor in different network capabilities. Any useful takeaways from the research?

**Comment by philrhc** [comment]
The paper is focused on container image distribution at the network edge so there are additional concerns like cloud connection availability and the vastly variable size of image layers.

Since container images are composed of layers of varying size, a request dispatcher component adds a three stage decision process:
1. It first queries the other nodes in the same LAN and fetches from the neighbor nodes. 
2. If not local, the dispatcher consults the manifest cache to obtain the layer’s size Li to determine if a P2P process shall be started. If Li is below a configurable threshold θ (e.g., 1 MiB), the request is routed directly to the upstream registry. This is justified by the data in Table II, which shows that a policy of direct fetch for small layers can handle a large fraction of requests with minimal latency.  
3. For layers larger than θ, the dispatcher initiates a P2P discovery process. However, this process is bounded by a dynamic timeout τ, defined as the 95th percentile of recent roundtrip times (RTTs) over a 10-second sliding window. If no suitable peer is found within τ , the system falls back to the upstream registry to avoid indefinite user wait times.

I think 1 is possible within torrent by prioritising peers with PeerSourceLPD in peer.go, those found by multicast. 
2 is interesting and I wanted to get your thoughts on @anacrolix

Another interesting takeaway of the paper is the addition of an autonomous tracker component that self-activates when no live trackers are reachable. it initiates a leader election within its local network partition using the FLOODMAX algorithm. In this protocol, each candidate broadcasts a stability score (based on metrics including uptime and known neighbors) and propagates the highest score observed so far. The node with the globally maximal score becomes the new tracker. While FLOODMAX ensures correctness, its naive implementation incurs O(n2) message complexity in dense networks. To address this, PeerSync applies path-pruning techniques that suppress redundant message forwarding across already-explored paths and avoid cross-subnet flooding unless necessary. This reduces communication overhead to nearlinear complexity while preserving convergence guarantees.

The last relevant takeaway is a proposal for block size based on layer size. 
Lb =
{  Li / 256,    if Li >= 1024 MiB
{  Li / 64,     if 256 MiB <= Li < 1024 MiB
{  Li / 16,     if 16 MiB <= Li < 256 MiB
{  Li,          otherwise

https://arxiv.org/abs/2507.20116

**Comment by anacrolix** [comment]
I still strongly support implementing this, might be a good vibe code opportunity.

**Comment by philrhc** [comment]
yes i agree, i would like to evaluate PeerSync against this library with this feature. 
How about a command line parameter like `-enableLocalPeerPrioritization=true`?

**Comment by anacrolix** [comment]
> yes i agree, i would like to evaluate PeerSync against this library with this feature. How about a command line parameter like `-enableLocalPeerPrioritization=true`?

Works for me!
